keystorage/src/main/java/tech/pegasys/web3signer/keystorage/awskms/AwsKeyManagerService.java

```diff
@@ -0,0 +1,172 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.web3signer.keystorage.awskms;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static tech.pegasys.web3signer.common.config.AwsAuthenticationMode.ENVIRONMENT;
+
+import tech.pegasys.web3signer.common.config.AwsAuthenticationMode;
+import tech.pegasys.web3signer.common.config.AwsCredentials;
+
+import java.net.URI;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.interfaces.ECPublicKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.Objects;
+import java.util.Optional;
+
+import com.google.common.annotations.VisibleForTesting;
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials;
+import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.core.SdkBytes;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.kms.KmsClient;
+import software.amazon.awssdk.services.kms.KmsClientBuilder;
+import software.amazon.awssdk.services.kms.model.GetPublicKeyRequest;
+import software.amazon.awssdk.services.kms.model.GetPublicKeyResponse;
+import software.amazon.awssdk.services.kms.model.KeySpec;
+import software.amazon.awssdk.services.kms.model.MessageType;
+import software.amazon.awssdk.services.kms.model.SignRequest;
+import software.amazon.awssdk.services.kms.model.SignResponse;
+import software.amazon.awssdk.services.kms.model.SigningAlgorithmSpec;
+
+/**
+ * Wraps AWS KMS Client and expose public key and sign operations. Since KmsClient is thread safe,
+ * same instance of this class can be used to talk to same host/credentials.
+ */
+public class AwsKeyManagerService implements AutoCloseable {
+  private final AwsAuthenticationMode authMode;
+  private final AwsCredentials awsCredentials;
+  private final String region;
+
+  private KmsClient kmsClient;
+
+  private Optional<URI> endpointOverride;
+
+  public AwsKeyManagerService(
+      final AwsAuthenticationMode authMode,
+      final AwsCredentials awsCredentials,
+      final String region,
+      final Optional<URI> endpointOverride) {
+    this.authMode = authMode;
+    this.awsCredentials = awsCredentials;
+    this.region = region;
+    this.endpointOverride = endpointOverride;
+
+    initKmsClient();
+  }
+
+  void initKmsClient() {
+    if (kmsClient != null) {
+      return;
+    }
```

Comment:
Is this thread-safe? What would be the consequence of constructing the client more than once for concurrent requests?

Reply:
Maybe we should use the Factory to maintain a singleton of this class?

Maybe related question: why is this class constructed from both AwsKMSSigner as well as AwsKMSSignerFactory?

Reply:
1. the `kmsClient != null` invariant is really not required as this initialization is meant to happen via constructor. 
2. We don't want this class to be a singelton. However, in future when we use bulk-loading, this class can be saved in a hashmap or a cache if needed, hence I implemented equals and hashcode methods.
3. In the AwsKMSSignerFactory, we need to obtain the ECPrimaryKey since that is required to construct Signer implementations (such as AzureSigner and AwsKMSSigner). The second construction happens in actual sign call (on-demand) as we need to call the sign method on it. Now that I re-review it, we might be able to avoid this construction by passing the same instance of `AwsKMS` inside `AwsKMSSigner`. (I adapted the logic from `AzureKeyVaultSigner` and `AzureKeyVaultSignerFactory`)